Zend/zend_compile.c

```diff
@@ -3054,6 +3056,19 @@ static zend_op *zend_delayed_compile_prop(znode *result, zend_ast *ast, uint32_t
 			}
 			zend_emit_jmp_null(&obj_node, type);
 		}
+
+		/* ZEND_AST_CONST can result in IS_TMP_VAR and IS_CONST which is uncommon for accessing
+		 * properties. ASSIGN_OBJ and FETCH_OBJ_[R]W are missing handlers for CONST|TMP op1, we work
+		 * around this by converting the TMP to a VAR, and by assigning the CONST to a temporary VAR. */
+		if ((type == BP_VAR_W || type == BP_VAR_RW)) {
+			if (obj_node.op_type == IS_TMP_VAR) {
+				ZEND_ASSERT(obj_ast->kind == ZEND_AST_CONST);
+				obj_node.op_type = IS_VAR;
```

Comment:
This means that ``FETCH_CONSTANT`` opline  has result_type IS_TMP_VAR and the following instruction use it as ``IS_VAR``.

```php
<?php
class Foo {
    public $array = [];
}
const FOO = new Foo();
FOO->array[] = 42;
```
```
0000 DECLARE_CONST string("FOO") zval(type=11)
0001 T1 = FETCH_CONSTANT string("FOO")
0002 V0 = FETCH_OBJ_W (dim write) V1 string("array")
0003 ASSIGN_DIM V0 NEXT
0004 OP_DATA int(42)
```
This may confuse some code in optimizer where we check if result and usage types are the same.
Can you also change the result_type of FETCH_CONSANT?

Reply:
Good catch. I wonder if at that point it's better to just adjust the VM handlers to `TMPVAR`, rather than convoluting the compiler. I suppose alternatively we could use a `QM_ASSIGN` here too, unless the optimizer removes it again.

Reply:
Do you see any problem in setting  FETCH_CONSANT.result_type to IS_TMP_VAR?

> I wonder if at that point it's better to just adjust the VM handlers to TMPVAR

This should introduce some overhead for all modified FETCH_..._W/RW or no?

Reply:
> Do you see any problem in setting FETCH_CONSANT.result_type to IS_TMP_VAR?

I don't see a problem. It might just introduce slightly more complexity in the compiler.

> This should introduce some overhead for all modified FETCH_..._W/RW or no?

I didn't think so, because we're changing them from `VAR` to `TMPVAR`. This should not actually change the specialized handlers. I will verify whether my assumption is correct. However, I will try your suggestion (changing result of FETCH_CONSTANT) first.

Reply:
I mean changing FETCH_CONSTANT.result_type only for this case.

```c
ZEND_ASSERT(opline->opcode == ZEND_FETCH_CONSTANT);
opline->result_type = IS_TMP_VAR;
```
May be my idea is wrong.